### PR TITLE
Fixes #15779 - make background processing available

### PR DESCRIPTION
**Pull request by ares**
blocked by https://github.com/theforeman/foreman-tasks/pull/194 but I no longer want to wait with publishing the PR so discussion/reviewing can start

### CODE REVIEW

**Comment by ares** [comment]
import
same works for importing puppet classes, recording stopped just before you'd see the same 'Update on background' submit button

report_import

**Comment by ares** [comment]
required version of foreman-tasks has been released so cancelling [Do not merge] flag

**Comment by ares** [comment]
[test]

**Comment by ohadlevy** [comment]
would it make sense to create a small progress bar (based on foreman tasks)?

I would remove the upgrade in the background, and always do it in the
background, but when watching the progress bar, you could click on leave
this page and continue in the background or something like that.

File: app/controllers/concerns/api/import_puppetclasses_common_controller.rb
Comment (iNecas) [inline app/controllers/concerns/api/import_puppetclasses_common_controller.rb]: You can use trigger_task https://github.com/theforeman/foreman-tasks/blob/master/lib/foreman_tasks.rb#L21, that takes the `async` as parameter and we would not need the `if/else` condition here

File: app/controllers/concerns/foreman/controller/environments.rb
Comment (iNecas) [inline app/controllers/concerns/foreman/controller/environments.rb]: Why not going the same approach as done in  `import_puppetclasses`?

**Comment by iNecas** [comment]
Would you mind writing tests for the actions, see https://github.com/theforeman/foreman_remote_execution/tree/master/test/unit/actions for inspiration

File: app/controllers/concerns/foreman/controller/environments.rb
Comment (ares) [reply app/controllers/concerns/foreman/controller/environments.rb]: Not entirely sure what you mean, could you please elaborate?

**Comment by ares** [comment]
> would it make sense to create a small progress bar (based on foreman tasks)?

that should probably belong to foreman-tasks, also I'm not sure what technology to use, react with data store is not there yet, maybe we could add simple patternfly based progress bar component with jquery that would update it through AJAX

> I would remove the upgrade in the background, and always do it in the
> background, but when watching the progress bar, you could click on leave
> this page and continue in the background or something like that.

would you display the progress bar through AJAX after button is clicked or would you redirect to some progress page? I guess foreman task detail would be too complicated for just watching progress

**Comment by ares** [comment]
> Would you mind writing tests for the actions, see https://github.com/theforeman/foreman_remote_execution/tree/master/test/unit/actions for inspiration

@iNecas what should the test cover? The only logic I see is constantization of a parameter and raising exception in case of some errors for report importing. Is that what you mean?

File: app/controllers/concerns/foreman/controller/environments.rb
Comment (iNecas) [reply app/controllers/concerns/foreman/controller/environments.rb]: My bad: I though it was an API controller. Makes sense

File: app/services/actions/foreman/puppet_class/import.rb
Comment (iNecas) [inline app/services/actions/foreman/puppet_class/import.rb]: If the reason was to show this in task details, defining `humanized_input` method is better way to do this

**Comment by iNecas** [comment]
@ares  app/services/actions/foreman/report/import.rb has some logic that could be tested. The main reason for this is, as introducing the tasks into foreman, to have also a test infrastrucuture ready for other cases

**Review by iNecas — APPROVED** [review]

File: app/services/actions/foreman/puppet_class/import.rb
Comment (iNecas) [inline app/services/actions/foreman/puppet_class/import.rb]: If the reason was to show this in task details, defining humanized_input method is better way to do this

**Review by iNecas — CHANGES_REQUESTED** [review]

File: Gemfile
Comment (mmoll) [inline Gemfile]: as a side note: for Debian this means, the foreman-tasks gem is included in the core package, conflicting with the current way of ruiby-foreman-tasks

Comment (iNecas) [reply Gemfile]: That would mean plugins currently using foreman-tasks would not need to specify them in the depedant gems anymore, right?

Comment (mmoll) [reply Gemfile]: Regarding Debian package dependencies, yes. It would also mean that -tasks and all its dependencies can't get updated seperately anymore on Debian without bumping the main foreman package..

Comment (ares) [reply Gemfile]: Seems we'd need some hack in https://github.com/theforeman/foreman-packaging/blob/deb/develop/debian/xenial/foreman/rules#L23. According to https://github.com/bundler/bundler/issues/2595 bundle package always need to get all gems. What if we removed it from vendor/cache and we'd keep it packaged as a separate dependency, probably moved from plugins repo to core?

Comment (mmoll) [reply Gemfile]: maybe before the whole bundler dance use sed to remove foreman-tasks from the Gemfile? But it doesn't feel right in any way ;)

File: app/services/actions/foreman/puppet_class/import.rb
Comment (ares) [reply app/services/actions/foreman/puppet_class/import.rb]: the reason was (sorry I could not remember yesterday) so that other tasks that might trigger this task could get the data and work with them further, not sure if this is needed or there's a better way

Comment (ares) [reply app/services/actions/foreman/puppet_class/import.rb]: oh, and also to avoid `{"results"=>[]}` default output I'm getting otherwise, not entirely sure where it comes from

**Comment by ares** [comment]
@iNecas I added test for report import action, could you please take another look?

File: Gemfile
Comment (ares) [reply Gemfile]: I agree, but we need to find a way to extract single deps from the bundle so we can move slowly migrate to properly packaged deps, otherwise it means we can never move from where we are

File: app/services/actions/foreman/puppet_class/import.rb
Comment (iNecas) [reply app/services/actions/foreman/puppet_class/import.rb]: One should be able to use the input from the task as well in additional tasks. The "results" thing looks strange, looking forward to look at it personally on Monday :)

**Comment by ares** [comment]
@iNecas I believe I changed what you requested, I also change the bookmark validator, hopefully it will fix the foreman tasks seed issue

**Comment by ares** [comment]
more tests appeared broken, changing to WoC again

**Comment by lzap** [comment]
I like you can choose if you want to process it in foreground and background. Maybe a drop down button with a default value (foreground?) could set a pattern we will use through the entire app.

**Comment by ares** [comment]
> I like you can choose if you want to process it in foreground and background. Maybe a drop down button with a default value (foreground?) could set a pattern we will use through the entire app.

I planned to do exactly what you've described :-) I'd like to create simple react component for that later (or someone could in parallel) but such helper probably belongs to foreman-tasks where this is not available yet and doing it using jquery would be probably rejected

**Comment by iNecas** [comment]
@ares I updated https://github.com/theforeman/foreman-tasks/pull/211 to also don't check the transactions for Concerns::ActionsTriggering, waiting for review there before we can proceed with releasing. I'm running tests locally for this PR to make sure it works, but looks good from code perspective

**Comment by iNecas** [comment]
Foreman tasks 0.8.4 and dynflow 0.8.17 with required changes pushed to rubygems

**Comment by ares** [comment]
I rebased the PR, let's look at Mr. Jenkins opinions

**Comment by iNecas** [comment]
I think I've found the reason for the failling tests https://github.com/theforeman/foreman-tasks/pull/214, jenkins seems to be much greener with it http://ci.theforeman.org/job/test_plugin_pull_request/3068/,

I've also noticed we will need to update the plugins tests job in jekins, https://github.com/theforeman/foreman-infra/blob/50a2d5b67fd18298cdd8b95414cdbbc9258e17cd/puppet/modules/jenkins_job_builder/files/theforeman.org/scripts/test/test_plugin.sh, to update the path to the foreman-tasks gem in Gemfile rather then adding the foreman-tasks as another gem, as bundler can't handle the gem being specified there twice

**Comment by iNecas** [comment]
Having said that, there are still some issues waiting for resolution, sqlite is green but mysql and postgres still seem to have some issue to be resolved

**Comment by iNecas** [comment]
Ok, so I have another candidate of source of troubles: the DatabaseCleaner truncation strategy, cleaning out dynflow_schema_info, this should help https://github.com/iNecas/foreman/commit/5d263d4f69e686eeea244d7507f85000ab39b06a trying out in http://ci.theforeman.org/job/test_plugin_pull_request/3070/

**Comment by iNecas** [comment]
Ok, so we are on 4 tests failing now http://ci.theforeman.org/job/test_plugin_pull_request/3070/database=postgresql,label=fast,ruby=2.1/#showFailuresLink that doesn't seem related to dynflow

File: Gemfile
Comment (mmoll) [reply Gemfile]: Could that dependency get externalized as `bundler.d/tasks.rb`? Then we have at least a chance to handle that different in packaging.

**Comment by ares** [comment]
rebased on @iNecas PR that should hopefully fix db migration for plugins

**Comment by ares** [comment]
I extracted the foreman_tasks gem from Gemfile, thanks @mmoll. @iNecas could you take one more look and approve changes if you have no more requirements?

**Review by iNecas — APPROVED** [review]

**Comment by iNecas** [comment]
ACK from my, switching to ready for merge and leaving time for next Wednesday, however, for further, non-blocker comments for enhancements, I would suggest leaving for separate issues/PRs

**Comment by iNecas** [comment]
Preparation for keeping plugin testing of foreman-tasks working is in https://github.com/theforeman/foreman-infra/pull/259

**Comment by iNecas** [comment]
installer changes https://github.com/theforeman/puppet-foreman/pull/504

File: app/controllers/concerns/foreman/controller/environments.rb
Comment (iNecas) [reply app/controllers/concerns/foreman/controller/environments.rb]: When doing final testing, I've noticed we include things like `authenticity` token in the input: I think we should limit the input only on the things we expect, which is `:changed` in this case

File: app/services/actions/foreman/puppet_class/import.rb
Comment (iNecas) [inline app/services/actions/foreman/puppet_class/import.rb]: The `changed[:new]` in my case was a strnig, not an array, resulting into the description, that that task has 170 clases vs. one:

```
changed:
  new:
    production: '{"motd":{"new":{"ensure":"present","config_file":"${$motd::params::config_file}","template":"${$motd::params::template}","inline_template":""}},"motd::params":{"new":{}}}'
commit: Update on background
controller: environments
action: obsolete_and_new
locale: 
```

File: app/services/actions/foreman/report/import.rb
Comment (iNecas) [inline app/services/actions/foreman/report/import.rb:15]:
```
params:
  config_report:
    host: tyan-gt24-09.rhts.eng.bos.redhat.com
```
`input[:params][:config_report]` should be used

File: app/controllers/api/v2/config_reports_controller.rb
Comment (iNecas) [inline app/controllers/api/v2/config_reports_controller.rb]: Similarly as with classes import, we should not push the whole params to the action, just the data we're interested in

**Review by iNecas — CHANGES_REQUESTED** [review]

File: app/services/actions/foreman/report/import.rb
Comment (ares) [reply app/services/actions/foreman/report/import.rb:15]: This was changed a while ago https://github.com/theforeman/puppet-foreman/pull/466 but the new version of module will be released with 1.14 I guess. For backwards compatibility, I'll look to :config_reports as fall back.

File: app/services/actions/foreman/puppet_class/import.rb
Comment (ares) [reply app/services/actions/foreman/puppet_class/import.rb]: Right, I also moved this to be humanized_output, since this input could be quite long. I fixed classes count to be correctly parsed and to be printed only when it make sense, e.g. when env is destroy we only get information that it was destroy with all of its classes so we don't print the classes number (would be always 1)

File: app/services/actions/foreman/report/import.rb
Comment (ares) [reply app/services/actions/foreman/report/import.rb:15]: fixed in config reports controller where I rename config_report to report (deprecated controller)

File: app/controllers/api/v2/config_reports_controller.rb
Comment (ares) [reply app/controllers/api/v2/config_reports_controller.rb]: fixed in both config reports and reports controller to only include report data

**Comment by ares** [comment]
@iNecas I addressed all you comments, for easier reviewing, see my [unsquashed branch](https://github.com/ares/foreman/tree/fix/15779_unsquashed) or the [actual commit](https://github.com/ares/foreman/commit/3ab93d9a92dd9d0ea286522806a6f41ce643a68b). I squashed and rebased everything in this branch.

**Review by iNecas — APPROVED** [review]

**Comment by iNecas** [comment]
Automated and manual tests passing, I don't see any point on waiting any longer, I hope this will unlock the whole load of things we were not able to do before. Merging

**Comment by iNecas** [comment]
thanks @ares

**Comment by iNecas** [comment]
RPM packaging PR is here https://github.com/theforeman/foreman-packaging/pull/1436

File: app/controllers/api/v2/config_reports_controller.rb
Comment (ekohl) [inline app/controllers/api/v2/config_reports_controller.rb:42]: A bit late to the party, but this should return a HTTP 202 Accepted. I don't know if currently does that.

File: app/controllers/api/v2/reports_controller.rb
Comment (ekohl) [inline app/controllers/api/v2/reports_controller.rb:42]: Should also be a HTTP 202.

File: app/controllers/api/v2/config_reports_controller.rb
Comment (ares) [reply app/controllers/api/v2/config_reports_controller.rb:42]: This comes through POST so we return 201 (:created), would you mind opening redmine issue related to #15779? I'm happy to address.

Comment (ekohl) [reply app/controllers/api/v2/config_reports_controller.rb:42]: http://projects.theforeman.org/issues/17483

**Comment by ohadlevy** [comment]
finally :+1:

**Pull request merged**
